### PR TITLE
New formula for Petsc complex to replace now missing option

### PR DESCRIPTION
**Pull request by izaid**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [X] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This is a continuation of https://github.com/Homebrew/homebrew-core/issues/29737.

Petsc must be compiled with either real or complex support. This amounts to simply changing `"--with-scalar-type=real"` to `"--with-scalar-type=complex"`, otherwise everything else is identical. There used to be an option to choose one of these in Brew, but it was taken away a few months ago. There are two potential ways forward:

1) Have a single formula that installs both versions of the library, with some way to select between them. I tried to create something like this, but was unsuccessful.

2) Have two formulae and the user can install either `petsc` (should it be `petsc-real`?) or `petsc-complex`. They conflict with each other.

This PR successfully implements 2). For whatever it is worth, Ubuntu provides two packages, with `update-alternatives` being used to select between them.

### CODE REVIEW

**Comment by izaid** [comment]
Pinging some relevant people for this, @ilovezfs @fmahebert

**Comment by izaid** [comment]
Bumping this. Happy to update or change something if need be.

**Comment by fmahebert** [comment]
Looks reasonable to me, but I'm not a Homebrew dev 🙂

File: Formula/petsc.rb [hunk 1]
```diff
@@ -17,22 +17,26 @@ class Petsc < Formula
...
   depends_on "scalapack"
   depends_on "suite-sparse"
+  conflicts_with "petsc-complex", :because => "petsc must be installed with either real or complex support, not both"
 
   def install
     ENV["CC"] = "mpicc"
     ENV["CXX"] = "mpicxx"
     ENV["F77"] = "mpif77"
     ENV["FC"] = "mpif90"
-    system "./configure", "--prefix=#{prefix}",
```
Comment (fxcoudert) [inline Formula/petsc.rb]: No unrelated changes to `petsc` please.

File: Formula/petsc.rb [hunk 2]
```diff
@@ -17,22 +17,26 @@ class Petsc < Formula
...
+      --with-x=0
+    ]
+    system "./configure", "--with-scalar-type=real", *args
     system "make", "all"
     system "make", "install"
   end
 
   test do
-    test_case = "#{pkgshare}/examples/ksp/ksp/examples/tutorials/ex1.c"
+    test_case = "#{share}/petsc/examples/ksp/ksp/examples/tutorials/ex1.c"
```
Comment (fxcoudert) [inline Formula/petsc.rb]: No unrelated changes to `petsc` please.

File: Formula/petsc-complex.rb [hunk 3]
```diff
@@ -0,0 +1,49 @@
+class PetscComplex < Formula
+  desc "Portable, Extensible Toolkit for Scientific Computation"
+  homepage "https://www.mcs.anl.gov/petsc/"
+  url "http://ftp.mcs.anl.gov/pub/petsc/release-snapshots/petsc-lite-3.9.3.tar.gz"
+  sha256 "8828fe1221f038d78a8eee3325cdb22ad1055a2f0671871815ee9f47365f93bb"
+
+  bottle do
```
Comment (fxcoudert) [inline Formula/petsc-complex.rb]: No bottle block

File: Formula/petsc-complex.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
+class PetscComplex < Formula
+  desc "Portable, Extensible Toolkit for Scientific Computation"
```
Comment (fxcoudert) [inline Formula/petsc-complex.rb]: Needs to state in the description that it is the complex version

File: Formula/petsc-complex.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+  end
+
+  depends_on "hdf5"
+  depends_on "hwloc"
+  depends_on "metis"
+  depends_on "netcdf"
+  depends_on "open-mpi"
+  depends_on "scalapack"
+  depends_on "suite-sparse"
+  conflicts_with "petsc", :because => "petsc must be installed with either real or complex support, not both"
```
Comment (fxcoudert) [inline Formula/petsc-complex.rb]: Blank line before `conflicts_with`

File: Formula/petsc-complex.rb [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+  def install
+    ENV["CC"] = "mpicc"
+    ENV["CXX"] = "mpicxx"
+    ENV["F77"] = "mpif77"
+    ENV["FC"] = "mpif90"
+
+    args = %W[
```
Comment (fxcoudert) [inline Formula/petsc-complex.rb]: No `args` variable, inline everything below.

File: Formula/petsc-complex.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      --with-debugging=0
+      --with-x=0
+    ]
+    system "./configure", "--with-scalar-type=complex", *args
+    system "make", "all"
+    system "make", "install"
+  end
+
+  test do
+    test_case = "#{share}/petsc/examples/ksp/ksp/examples/tutorials/ex1.c"
```
Comment (fxcoudert) [inline Formula/petsc-complex.rb]: Use `pkgshare`

**Review by fxcoudert — CHANGES_REQUESTED** [review]

**Comment by izaid** [comment]
@fxcoudert Thanks very much for the comments! I have made all of the changes you requested, except for one (see below). 

For the `petsc-complex` formula specifically, the `pkgshare` variable can't be used because `petsc-complex` is installed with the same directory as `petsc`. This is important so that the C include directory remains the same, regardless of whether one is using `petsc` or `petsc-complex`. That's why the `share` variable is used there.

If there are any other changes you want, just ping me. Otherwise, would be happy to get this merged in.

File: Formula/petsc-complex.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
+class PetscComplex < Formula
+  desc "Portable, Extensible Toolkit for Scientific Computation (with complex support)"
```
Comment (zbeekman) [inline Formula/petsc-complex.rb]: TestBot is failing due to brew audit; the formula description is 93 chars but needs to be ,= 80. Maybe instead of `(with complex support)` we could say `(complex)`? Anyone else have thoughts on how to appease brew audit?

**Review by zbeekman — DISMISSED** [review]

File: Formula/petsc-complex.rb (hunk 7 above)
Comment (zbeekman) [reply Formula/petsc-complex.rb]: @izaid said:
> For the petsc-complex formula specifically, the pkgshare variable can't be used because petsc-complex is installed with the same directory as petsc. This is important so that the C include directory remains the same, regardless of whether one is using petsc or petsc-complex. That's why the share variable is used there.

But you're setting `--prefix=#{prefix}`. Am I missing something? Doesn't petsc go into e.g., `/usr/local/Cellar/petsc` and petsc-complex go into `/usr/local/Cellar/petsc-complex`?

**Comment by izaid** [comment]
@zbeekman I have now used `(complex)` in the description of the complex formula, and correspondingly added `(real)` in the description of the real formula. Thought it better to be explicit in the real formula as well.

**Comment by zbeekman** [comment]
@izaid can you please address my comment above RE `--prefix=#{prefix}` and `pkgshare` vs `share`?

Thanks.

File: Formula/petsc-complex.rb (hunk 7 above)
Comment (izaid) [reply Formula/petsc-complex.rb]: I just printed out the variables. Their values are as follows

`prefix = /usr/local/Cellar/petsc-complex/3.9.3`
`share = /usr/local/Cellar/petsc-complex/3.9.3/share`
 `pkgshare = /usr/local/Cellar/petsc-complex/3.9.3/share/petsc-complex`

So `prefix` and `share` are fine. However, `petsc` puts things into `/usr/local/Cellar/petsc-complex/3.9.3/share/petsc`. Consider that for include, i.e. `/usr/local/Cellar/petsc-complex/3.9.3/include/petsc`, this makes sense, so that the C include directory doesn't change depending upon whether or not `petsc` or `petsc-complex` is installed.

**Comment by izaid** [comment]
@zbeekman Sorry, I didn't see your comment yesterday. Just addressed it.

**Review by fxcoudert — APPROVED** [review]

**Comment by fxcoudert** [comment]
Thanks @izaid for your contribution to Homebrew! This is now merged
